Mathlib/RingTheory/SurjectiveOnStalks.lean

```diff
@@ -0,0 +1,185 @@
+    obtain ⟨c, hc⟩ := ha
+    simp only [← mul_assoc] at hc
+    exact ⟨_, _, _, c.2, t.2, hc.symm⟩
+  · refine fun H y ↦ Localization.ind (fun ⟨y, t, h⟩ ↦ ?_) y
+    simp only
+    obtain ⟨yx, ys, yc, hyc, hy, ey⟩ := H y
+    obtain ⟨tx, ts, yt, hyt, ht, et⟩ := H t
+    refine ⟨Localization.mk (yx * ts) ⟨ys * tx, Submonoid.mul_mem _ hy ?_⟩, ?_⟩
+    · exact fun H ↦ mul_mem (P.primeCompl.mul_mem hyt ht) h (et ▸ Ideal.mul_mem_left _ yt H)
+    · simp only [Localization.mk_eq_mk', Localization.localRingHom_mk', map_mul f,
+        IsLocalization.mk'_eq_iff_eq, IsLocalization.eq_iff_exists P.primeCompl]
+      refine ⟨⟨yc, hyc⟩ * ⟨yt, hyt⟩, ?_⟩
+      simp only [Submonoid.coe_mul]
+      convert congr($(ey.symm) * $(et)) using 1 <;> ring
+
+lemma surjectiveOnStalks_iff_forall_ideal :
+    f.SurjectiveOnStalks ↔
+      ∀ I : Ideal S, I ≠ ⊤ → ∀ s : S, ∃ x r : R, ∃ c ∉ I, f r ∉ I ∧ c * f r * s = c * f x := by
+  simp_rw [SurjectiveOnStalks, surjective_localRingHom_iff]
+  refine ⟨fun H I hI s ↦ ?_, fun H I hI ↦ H I hI.ne_top⟩
+  obtain ⟨M, hM, hIM⟩ := I.exists_le_maximal hI
+  obtain ⟨x, r, c, hc, hr, e⟩ := H M hM.isPrime s
+  exact ⟨x, r, c, fun h ↦ hc (hIM h), fun h ↦ hr (hIM h), e⟩
+
+lemma surjectiveOnStalks_iff_forall_maximal :
+    f.SurjectiveOnStalks ↔ ∀ (I : Ideal S) (_ : I.IsMaximal),
+      Function.Surjective (Localization.localRingHom _ I f rfl) := by
+  refine ⟨fun H I hI ↦ H I hI.isPrime, fun H I hI ↦ ?_⟩
+  simp_rw [surjective_localRingHom_iff] at H ⊢
+  intro s
+  obtain ⟨M, hM, hIM⟩ := I.exists_le_maximal hI.ne_top
+  obtain ⟨x, r, c, hc, hr, e⟩ := H M hM s
+  exact ⟨x, r, c, fun h ↦ hc (hIM h), fun h ↦ hr (hIM h), e⟩
+
+lemma surjectiveOnStalks_iff_forall_maximal' :
+    f.SurjectiveOnStalks ↔ ∀ I : Ideal S, I.IsMaximal →
+      ∀ s : S, ∃ x r : R, ∃ c ∉ I, f r ∉ I ∧ c * f r * s = c * f x := by
+  simp only [surjectiveOnStalks_iff_forall_maximal, surjective_localRingHom_iff]
+
+lemma surjectiveOnStalks_of_exists_div (h : ∀ x : S, ∃ r s : R, IsUnit (f s) ∧ f s * x = f r) :
+    SurjectiveOnStalks f :=
+  surjectiveOnStalks_iff_forall_ideal.mpr fun I hI x ↦
+    let ⟨r, s, hr, hr'⟩ := h x
+    ⟨r, s, 1, by simpa [← Ideal.eq_top_iff_one], fun h ↦ hI (I.eq_top_of_isUnit_mem h hr), by simpa⟩
+
+lemma surjectiveOnStalks_of_surjective (h : Function.Surjective f) :
+    SurjectiveOnStalks f :=
+  surjectiveOnStalks_iff_forall_ideal.mpr fun _ _ s ↦
+    let ⟨r, hr⟩ := h s
+    ⟨r, 1, 1, by simpa [← Ideal.eq_top_iff_one], by simpa [← Ideal.eq_top_iff_one], by simp [hr]⟩
+
+variable (S) in
+lemma surjectiveOnStalks_of_isLocalization [IsLocalization M S] :
+    SurjectiveOnStalks (algebraMap R S) := by
+  refine surjectiveOnStalks_of_exists_div fun s ↦ ?_
+  obtain ⟨x, s, rfl⟩ := IsLocalization.mk'_surjective M s
+  exact ⟨x, s, IsLocalization.map_units S s, IsLocalization.mk'_spec' S x s⟩
+
+lemma SurjectiveOnStalks.comp (hg : SurjectiveOnStalks g) (hf : SurjectiveOnStalks f) :
```

Comment:
Besides composition, is it relevant to add the case of `id` or would the case of localization take care of it?

Reply:
`surjectiveOnStalks_of_surjective` should cover it quite easily.